src/StreamJsonRpc/JsonRpc.cs

```diff
@@ -946,11 +960,34 @@ private JsonRpcMessage HandleInvocationTaskResult(JToken id, Task t)
 
             if (t.IsFaulted)
             {
+                if (t.Exception != null)
+                {
+                    if (t.Exception.InnerException != null)
+                    {
+                        if (this.IsFatalException(t.Exception.InnerException))
+                        {
+                            ExceptionDispatchInfo.Capture(t.Exception.InnerException).Throw();
+                        }
+                    }
+                    else
+                    {
+                        if (this.IsFatalException(t.Exception))
+                        {
+                            ExceptionDispatchInfo.Capture(t.Exception).Throw();
+                        }
+                    }
+                }
+
                 return CreateError(id, t.Exception);
             }
 
             if (t.IsCanceled)
             {
+                if (this.IsFatalException(new OperationCanceledException()))
```

Comment:
I don't think we should manufacture an exception here. If the method we invoked doesn't actually throw an exception, but rather returns a Canceled task, that likely means that the server method went out of its way to skip an exception code path and we shouldn't artificially bring it back. Anyway, the method is called `IsFatalException` and no exception occurred.

Reply:
I found that any time an asynchronous task was canceled (or if the server method threw `OperationCanceledException` it ended up in this handler as a canceled task. I agree though, this felt weird, and especially with there being no way to distinguish between an exception thrown or a canceled task returned.